src/cat.jl

```diff
@@ -0,0 +1,69 @@
+
+################################################
+# cat, vcat, hcat
+
+function Base.cat(a::NamedDimsArray{L}; dims) where L
+    newL = expand_dimnames(L, dims)
+    numerical_dims = dim(newL, dims)
+    data = Base.cat(parent(a); dims=numerical_dims)
+    newL = unify_names_longest(newL, dimnames(data))
+    return NamedDimsArray{newL}(data)
+end
+
+function Base.cat(a::NamedDimsArray{L}, b::AbstractArray; dims) where L
+    newL = expand_dimnames(L, dims)
+    numerical_dims = dim(newL, dims)
+    data = Base.cat(parent(a), b; dims=numerical_dims)
+    newL = unify_names_longest(newL, dimnames(data)) # when dims=3 for two 2d arrays
+    return NamedDimsArray{newL}(data)
+end
+
+function Base.cat(a::AbstractArray, b::NamedDimsArray{L}; dims) where L
+    newL = expand_dimnames(L, dims)
+    numerical_dims = dim(newL, dims)
+    data = Base.cat(a, parent(b); dims=numerical_dims)
+    newL = unify_names_longest(newL, dimnames(data)) # when dims=3 for two 2d arrays
+    return NamedDimsArray{newL}(data)
+end
+
+function Base.cat(a::NamedDimsArray{La}, b::NamedDimsArray{Lb}; dims) where {La, Lb}
+    newL = unify_names_shortest(La, Lb)
+    newL = expand_dimnames(newL, dims)
+    numerical_dims = dim(newL, dims)
+    data = Base.cat(parent(a), parent(b); dims=numerical_dims)
+    newL = unify_names_longest(newL, dimnames(data)) # when dims=3 for two 2d arrays
+    return NamedDimsArray{newL}(data)
+end
+
+# to dispatch on the first or the second argument being the NDA
+for (T, S) in [
+    (:NamedDimsArray, :AbstractArray),
+    (:AbstractArray, :NamedDimsArray),
+    (:NamedDimsArray, :NamedDimsArray)
+    ]
+
+    @eval function Base.cat(a::$T, b::$S, c::AbstractArray...; dims)
+        return Base.cat(a, Base.cat(b, Base.cat(c...; dims=dims); dims=dims); dims=dims)
+    end
+end
+
+# Base.hcat and Base.vcat specialise on this Union
+const AbsVecOrMat = Union{AbstractVector, AbstractMatrix}
+for (T, S) in [
+    (:NamedDimsArray, :AbsVecOrMat),
+    (:NamedDimsArray, :AbstractArray),
+    (:AbsVecOrMat, :NamedDimsArray),
+    (:AbstractArray, :NamedDimsArray),
+    (:NamedDimsArray, :NamedDimsArray)
+    ]
+
+    for (fun, d) in zip((:vcat, :hcat), (1, 2))
+        @eval function Base.$fun(a::$T, b::$S)
+            return Base.cat(a, b; dims=$d)
+        end
+
+        @eval function Base.$fun(a::$T, b::$S, c::NamedDimsArray...)
+            return Base.cat(a, b, c...; dims=$d)
```

Comment:
I don't think `hcat` should call `cat`, despite giving the same result:
```julia
julia> @btime hcat(A,B) setup=(A=rand(2,2); B=rand(2,2));
  54.956 ns (2 allocations: 176 bytes)

julia> _hcat(xs...) = cat(xs...; dims=2);

julia> @btime _hcat(A,B) setup=(A=rand(2,2); B=rand(2,2));
  3.812 μs (30 allocations: 1.36 KiB)
```
Also I think you want `c::AbstractArray...` as above; and this method should also cover the 2-arg case.

Reply:
Wow, that's a huge difference. I guess the alternative is to implement it in the same way as cat, i.e. calling hcat with parent arrays?

Right, I agree, `c::AbstractArray` was what I tried first but got method ambiguities. Will see if it can work this way with the alternative implementation.

I thought splatting `c` means that 2-arg case is also covered? i.e.
```
julia> function myfunc(a::Int, b::Int, c::Float64...)
           return 0
       end

julia> myfunc(3, 3, 3.0)
0

julia> myfunc(3, 3)
0

```

Reply:
Ambiguities are a pain, I didn't check. But yes, re "splatting c means that 2-arg case is also covered" that's what I was trying to say too. 

I think things like `cat(a, cat(b, cat(c...;` will also be very slow. Maybe `unify_names_longest` needs a vararg method, like `unify_names` has (which is not slow because it's all tuples).

Reply:
Ah, yes, I understand your comment now - I had to add the separate 2-arg methods because of ambiguities as well.

Reply:
Okay, all done now. Had to replace the my much beloved tuple comprehension (yay Python) with `ntuple(i->dimnames(c[i]), length(c))`.

```
julia> @ballocated (()->cat(a, a, a, dims=1))()
1824
julia> @ballocated (()->cat(nda, nda, nda, dims=1))()
1824

julia> @ballocated (()->vcat(nda, nda, nda))()
176
julia> @ballocated (()->vcat(a, a, a))()
176
```

Thank you @mcabbott for flagging this!